File: typescript/src/resources/signing-keys.ts
/**
 * Org-level webhook signing key management.
 *
 * Shared across all Inkbox clients (mail, phone, etc.).
 */

import { HttpTransport } from "../_http.js";

const PATH = "/signing-keys";

export interface SigningKey {
  /** Plaintext signing key — returned once on creation/rotation. Store securely. */
  signingKey: string;
  createdAt: Date;
}

interface RawSigningKey {
  signing_key: string;
  created_at: string;
}

function parseSigningKey(r: RawSigningKey): SigningKey {
  return {
    signingKey: r.signing_key,
    createdAt: new Date(r.created_at),
  };
}

export class SigningKeysResource {
  constructor(private readonly http: HttpTransport) {}

  /**
   * Create or rotate the webhook signing key for your organisation.
   *
   * The first call creates a new key; subsequent calls rotate (replace) the
   * existing key. The plaintext `signingKey` is returned **once** —
   * store it securely as it cannot be retrieved again.
   *
   * Use the returned key to verify `X-Inkbox-Signature` headers on
   * incoming webhook requests.
   */
  async createOrRotate(): Promise<SigningKey> {
    const data = await this.http.post<RawSigningKey>(PATH, {});
    return parseSigningKey(data);
  }
}

File: python/inkbox/signing_keys.py
"""
inkbox/signing_keys.py

Org-level webhook signing key management — shared across all Inkbox clients.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Any


@dataclass
class SigningKey:
    """Org-level webhook signing key.

    Returned once on creation/rotation — store ``signing_key`` securely.
    """

    signing_key: str
    created_at: datetime

    @classmethod
    def _from_dict(cls, d: dict[str, Any]) -> SigningKey:
        return cls(
            signing_key=d["signing_key"],
            created_at=datetime.fromisoformat(d["created_at"]),
        )


class SigningKeysResource:
    def __init__(self, http: Any) -> None:
        self._http = http

    def create_or_rotate(self) -> SigningKey:
        """Create or rotate the webhook signing key for your organisation.

        The first call creates a new key; subsequent calls rotate (replace) the
        existing key. The plaintext ``signing_key`` is returned **once** —
        store it securely as it cannot be retrieved again.

        Use the returned key to verify ``X-Inkbox-Signature`` headers on
        incoming webhook requests.

        Returns:
            The newly created/rotated signing key with its creation timestamp.
        """
        data = self._http.post("/signing-keys", json={})
        return SigningKey._from_dict(data)
